packages/SwingSet/src/kernel/vatManager/manager-local.js

```diff
@@ -129,43 +151,26 @@ export function makeLocalVatManagerFactory(tools) {
       enableVatstore,
       gcTools,
       makeVatConsole(makeLogMaker(liveSlotsConsole)),
+      buildVatNamespace,
     );
 
-    const endowments = harden({
-      ...vatEndowments,
-      ...ls.vatGlobals,
-      console: makeVatConsole(makeLogMaker(vatConsole)),
-      assert,
-      TextEncoder,
-      TextDecoder,
-      Base64: globalThis.Base64, // Available only on XSnap
-      URL: globalThis.URL, // Unavailable only on XSnap
-    });
-    const inescapableGlobalProperties = { ...ls.inescapableGlobalProperties };
-
-    const vatNS = await importBundle(bundle, {
-      filePrefix: `vat-${vatID}/...`,
-      endowments,
-      inescapableGlobalProperties,
-    });
-
-    let dispatch;
-    if (typeof vatNS.buildRootObject === 'function') {
-      const { buildRootObject } = vatNS;
-      ls.setBuildRootObject(buildRootObject);
-      dispatch = ls.dispatch;
-    } else if (enableSetup) {
+    if (enableSetup) {
+      const vatNS = await buildVatNamespace(
+        ls.vatGlobals,
+        ls.inescapableGlobalProperties,
+      );
       const setup = vatNS.default;
       assert(setup, X`vat source bundle lacks (default) setup() function`);
       assert.typeof(setup, 'function');
       const helpers = harden({}); // DEPRECATED, todo remove from setup()
       const state = null; // TODO remove from setup()
-      dispatch = setup(syscall, state, helpers, vatPowers, vatParameters);
+      const dispatch = setup(syscall, state, helpers, vatPowers, vatParameters);
+      return finish(dispatch);
     } else {
-      assert.fail(X`vat source bundle lacks buildRootObject() function`);
+      assert(ls.dispatch);
+      const result = finish(ls.dispatch);
```

Comment:
Yeah, I thought about that.  I really wish the JS `const` declaration simply enforced single assignment semantics rather than requiring immediate initialization.  As it is, to do what you say we'd have to declare `dispatch` as a `let` variable outside the `if`, and I'm always nervous about that kind of thing, even though, in this case, it's probably not an issue since the value is going to immediately be returned.